File: src/counter_explore.c
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <unistd.h>

#include "pr_utils.h" 
#include "cce_1_def.h" 

/*
		gcc -o adder adder.c  -lpigpio -lpthread
	
*/

int My_logic_function ( int* inpA , int* inpB , int insize, int*  outpA, int* outpB, int outsize  ) {
	if ( insize > 8 ) return -1; 
	if ( outsize > 8 ) return -2; 
	for ( int inpcnt=0; inpcnt < insize ; inpcnt++) {
			inpA[inpcnt] = inpA[inpcnt] & 1;
			inpB[inpcnt] = inpB[inpcnt] & 1;
	}
	int prevcarry=0; //dummy input carry for the adder 
	for ( int nr = 0 ; nr < 8 ; nr++) {
		// define here a full adder  nr is the follow nr of the adder 
		outpA[nr] = (inpA[nr] ^ inpB[nr]) ^ prevcarry ;  // sum is xor of the inputs of a full addr 
		prevcarry=(inpA[nr] & inpB[nr]) | (( (inpA[nr] ^ inpB[nr])) & prevcarry);
		outpB[nr]=prevcarry;
	}
	return 0;
} 
	


int main(int argc, char *argv[]){
// --- this is setup the hardware for the test 
	int err= init_cce_1();
	if (err) {
		printf("initialization of hardware failed with err %d \n\r", err);
		return err; 
	}
	err=set_muxout( 'B' );  // mux_8_4 has tobe in position B to get the counters 
	if (err) {
		printf("can not set the output mux correctly for this test  err %d \n\r", err);
		return err; 
	}
// define test specific input / outputs
	// now define my inputs 
	const int nr_inputs=4;
	unsigned  inputs[]= {D0,D1,D2,D3} ; // preset for the counters inputs for the counter
	if (sizeof(inputs) / sizeof(inputs[0]) < nr_inputs) { 
		printf("input pin array badly defined \n\r");
		return -nr_inputs ;
	}
// these drive the inputs of the FPGA so have to defined as outputs 
	for ( int inpcnt=0; inpcnt < nr_inputs; inpcnt++) {
		err= gpioSetMode( inputs[inpcnt],PI_OUTPUT);
		if( err) { printf("can not set gpio %d to output with err %d \n\r",inputs[inpcnt], err); 
			return err;
		}
	}
    // special  signals 
	const int DSEL = D5 ; // is a output for loading a preset value 
	gpioSetMode( DSEL , PI_OUTPUT);
	gpioWrite (DSEL, 0) ;	
	const int DCE= D4;  // count enable signal
	gpioSetMode( DCE  , PI_OUTPUT);
	gpioWrite (DCE, 1) ;	 // has to be active
    // the clock is an output for the RP 
	gpioSetMode( clk1 , PI_OUTPUT);
	gpioWrite (clk1, 0) ;	
	// the mux control  for the mux4_2
	const int m42SEL= GPIO22 ;
	gpioSetMode( m42SEL  , PI_OUTPUT);
	gpioWrite (m42SEL, 1) ;	

	set_outputs( inputs, nr_inputs , 4) ;	
	const int nr_status=8; 
	int bin_outp_arry[nr_status]; // result array 
// -- here start the test 	
	for (int cnt =0;cnt < 20;cnt++){
		gpioWrite (clk1, 0) ;	
		printf ("cnt = %3d , clk= %d ",cnt, gpioRead(clk1));
		print_led_status (2);
		//sleep(1);
		gpioWrite (clk1, 1) ;	
		printf ("cnt = %3d , clk= %d ",cnt, gpioRead(clk1));
		print_led_status (2);
		sleep(1);
	}
	
	
	 gpioTerminate();
	return 0;  // will not be reached. 
}

File: python/counter_explore.py
import time
import pigpio
from upctrl_pinfunctions import *
import pr_utils 

def My_logic_function(inpA, inpB, insize, outpA, outpB, outsize):
    if insize > 8:
        return -1
    if outsize > 8:
        return -2
    for inpcnt in range(insize):
        inpA[inpcnt] = inpA[inpcnt] & 1
        inpB[inpcnt] = inpB[inpcnt] & 1
    prevcarry = 0
    for nr in range(8):
        outpA[nr] = (inpA[nr] ^ inpB[nr]) ^ prevcarry
        prevcarry = (inpA[nr] & inpB[nr]) | ((inpA[nr] ^ inpB[nr]) & prevcarry)
        outpB[nr] = prevcarry
    return 0

def main():
    
    print("start program ", {__file__})
    err = init_cce_1()
    if err:
       print("initialization of hardware failed with err ", err)
       exit(err)
    set_muxout('B')  # set the routing of the signals on the FPGA board
    
    #special  signals 
    DSEL = D5 #is an output for loading a preset value 
    #Selector for the Multiplexer: The input of the flop-flop will be 
    #either the inverted output of the flip-flop (if DSEL==1), or a starting value for the counter given by D0-D3 (if DSEL==0).
    DCE= D4
    m42SEL= GPIO22 
    pi.set_mode(DSEL, pigpio.OUTPUT)
    pi.set_mode(DCE, pigpio.OUTPUT)
    pi.set_mode(clk1, pigpio.OUTPUT)
    pi.set_mode(m42SEL, pigpio.OUTPUT)
    
    pi.write(DSEL, 0)
    pi.write(DCE, 1)
    pi.write(clk1, 0)
    pi.write(m42SEL, 1)
    
    inputs = [D0, D1, D2, D3]
    for inpcnt in range(len(inputs)):
        pi.set_mode(inputs[inpcnt], pigpio.OUTPUT)
    
    bin_outp_arry = [0] * 8
    
    for cnt in range(20):
        pi.write(clk1, 0)
        print(f"cnt = {cnt}, clk = {pi.read(clk1)}")
        # Assuming print_led_status(2) is a function that prints LED status
        print_led_status(2)
        time.sleep(1)
        
        pi.write(clk1, 1)
        print(f"cnt_= {cnt}, clk = {pi.read(clk1)}")
        print_led_status(2)
        
        time.sleep(1)
    
    pi.stop()

if __name__ == "__main__":
    main()

